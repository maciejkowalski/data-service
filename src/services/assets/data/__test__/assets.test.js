const Task = require('folktale/concurrency/task');

const Maybe = require('folktale/maybe');
const { driver } = require('../../../../db/__test__/mocks/index');

const create = require('../create');

afterEach(() => jest.clearAllMocks());

describe('Asset methods should correctly', () => {
  const dbManyResp = () => [1, 2, 3];
  const dbOneResp = () => [1];

  const deps = {
    batchQuery: () => x => x,
    sql: jest.fn(x => x),
  };
  // const dataBad = create({ ...deps, pg: driverTBad });

  // const goodAdapter = adapter.good(headSecondArg);
  // const badAdapter = adapter.bad(headSecondArg);

  it('resolve many', done => {
    const pg = driver.create(Task.of, dbManyResp);
    const pgAnySpy = jest.spyOn(pg, 'any');
    const data = create({ ...deps, pg });

    data
      .mget([1, 2, 3])
      .run()
      .listen({
        onResolved: xs => {
          expect(xs).toMatchSnapshot();
          expect(deps.sql).toHaveBeenCalledWith([1, 2, 3]);
          expect(pgAnySpy).toHaveBeenCalledWith([1, 2, 3]);
          done();
        },
      });
  });

  it('reject many', done => {
    const pg = driver.create(Task.rejected, dbManyResp);
    const pgAnySpy = jest.spyOn(pg, 'any');
    const data = create({ ...deps, pg });

    data
      .mget([1, 2, 3])
      .run()
      .listen({
        onRejected: e => {
          expect(e).toMatchSnapshot();
          expect(deps.sql).toHaveBeenCalledWith([1, 2, 3]);
          expect(pgAnySpy).toHaveBeenCalledWith([1, 2, 3]);
          done();
        },
      });
  });

  it('resolve one', done => {
    const pg = driver.create(Task.of, dbOneResp);
    const pgAnySpy = jest.spyOn(pg, 'any');
    const data = create({ ...deps, pg });

    data
      .get(1)
      .run()
      .listen({
        onResolved: xs => {
          expect(xs).toMatchSnapshot();
          expect(deps.sql).toHaveBeenCalledWith([1]);
          expect(pgAnySpy).toHaveBeenCalledWith([1]);
          done();
        },
      });
  });

  it('reject one', done => {
    const pg = driver.create(Task.rejected, dbOneResp);
    const pgAnySpy = jest.spyOn(pg, 'any');
    const data = create({ ...deps, pg });

    data
      .get(1)
      .run()
      .listen({
        onRejected: e => {
          expect(e).toMatchSnapshot();
          expect(deps.sql).toHaveBeenCalledWith([1]);
          expect(pgAnySpy).toHaveBeenCalledWith([1]);
          done();
        },
      });
  });
});
