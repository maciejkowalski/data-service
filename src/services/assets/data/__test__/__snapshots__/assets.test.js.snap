// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset methods should correctly reject many 1`] = `
Object {
  "error": [Error: 1,2,3],
  "meta": Object {
    "params": Array [
      1,
      2,
      3,
    ],
    "request": "assets",
  },
  "type": "DBError",
}
`;

exports[`Asset methods should correctly reject one 1`] = `
Object {
  "error": [Error: 1],
  "meta": Object {
    "params": Array [
      1,
    ],
    "request": "assets",
  },
  "type": "DBError",
}
`;

exports[`Asset methods should correctly resolve many 1`] = `
Array [
  Object {
    "@@tag": "Just",
    "@@type": "folktale:Maybe",
    "@@value": Object {
      "value": 1,
    },
  },
  Object {
    "@@tag": "Just",
    "@@type": "folktale:Maybe",
    "@@value": Object {
      "value": 2,
    },
  },
  Object {
    "@@tag": "Just",
    "@@type": "folktale:Maybe",
    "@@value": Object {
      "value": 3,
    },
  },
]
`;

exports[`Asset methods should correctly resolve one 1`] = `
Object {
  "@@tag": "Just",
  "@@type": "folktale:Maybe",
  "@@value": Object {
    "value": 1,
  },
}
`;
